### PR TITLE
feat: "spark-submit jar on k8s" related label and resource expansion

### PR DESCRIPTION
**Pull request by Zhao-LX2000**
### What is the purpose of the change

Enable k8s cluster recognition and k8s resource management.

### Brief change log
- Expand UniversalCmdTemplate for recognizing "-k8sCluster";
- Modify ClusterLabel extension;
- Add k8s resource related classes.

### Checklist

- [x] I have read the [Contributing Guidelines on pull requests](https://github.com/facebook/docusaurus/blob/main/CONTRIBUTING.md#pull-requests).
- [x] I have explained the need for this PR and the problem it solves
- [x] I have explained the changes or the new features added to this PR
- [ ] I have added tests corresponding to this change
- [ ] I have updated the documentation to reflect this change
- [ ] I have verified that this change is backward compatible (If not, please discuss on the [Linkis mailing list](https://linkis.apache.org/community/how-to-subscribe) first)
- [ ] **If this is a code change**: I have written unit tests to fully verify the new behavior.

### CODE REVIEW

**Comment by aiceflower** [comment]
Please handle the third dependency check problem.

**Comment by Zhao-LX2000** [comment]
> Please handle the third dependency check problem.

Thanks. Already fix and please check.

**Comment by peacewong** [comment]
@guoshupei

File: linkis-computation-governance/linkis-manager/linkis-label-common/src/main/java/org/apache/linkis/manager/label/entity/cluster/ClusterLabel.java [hunk 1]
```diff
@@ -18,14 +18,17 @@
...
 
 import java.util.HashMap;
 
-public class ClusterLabel extends GenericLabel implements EMNodeLabel {
+import static org.apache.linkis.manager.label.errorcode.LabelCommonErrorCodeSummary.LABEL_ERROR_CODE;
+
+public class ClusterLabel extends GenericLabel implements EngineNodeLabel, UserModifiable {
 
   public ClusterLabel() {
     setLabelKey(LabelKeyConstant.YARN_CLUSTER_KEY);
```
Comment (guoshupei) [inline linkis-computation-governance/linkis-manager/linkis-label-common/src/main/java/org/apache/linkis/manager/label/entity/cluster/ClusterLabel.java:34]: Whether the `LabelKey` can be changed

File: linkis-computation-governance/linkis-manager/linkis-application-manager/src/main/java/org/apache/linkis/manager/rm/external/kubernetes/KubernetesResourceRequester.java [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+import io.fabric8.kubernetes.api.model.metrics.v1beta1.NodeMetrics;
+import io.fabric8.kubernetes.client.ConfigBuilder;
+import io.fabric8.kubernetes.client.DefaultKubernetesClient;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class KubernetesResourceRequester implements ExternalResourceRequester {
+  private static final Logger logger = LoggerFactory.getLogger(KubernetesResourceRequester.class);
+  private ExternalResourceProvider provider = null;
+  private String k8sMasterUrl;
```
Comment (guoshupei) [inline linkis-computation-governance/linkis-manager/linkis-application-manager/src/main/java/org/apache/linkis/manager/rm/external/kubernetes/KubernetesResourceRequester.java:38]: How to support different namespaces and different queues?

File: linkis-computation-governance/linkis-manager/linkis-application-manager/src/main/java/org/apache/linkis/manager/rm/external/kubernetes/KubernetesResourceRequester.java [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+    long usedCPU = 0;
+    long allocatableCPU = 0;
+    for (NodeMetrics item : client.top().nodes().metrics().getItems()) {
+      usedMemory += Long.parseLong(item.getUsage().get("memory").getAmount());
+      usedCPU += Long.parseLong(item.getUsage().get("cpu").getAmount()) / 1000000L;
+    }
+    for (Node item : client.nodes().list().getItems()) {
+      allocatableMemory +=
+          Long.parseLong(item.getStatus().getAllocatable().get("memory").getAmount());
+      allocatableCPU += Long.parseLong(item.getStatus().getAllocatable().get("cpu").getAmount());
```
Comment (guoshupei) [inline linkis-computation-governance/linkis-manager/linkis-application-manager/src/main/java/org/apache/linkis/manager/rm/external/kubernetes/KubernetesResourceRequester.java:70]: Whether the reference `usedCPU` is correct?

File: linkis-computation-governance/linkis-manager/linkis-label-common/src/main/java/org/apache/linkis/manager/label/entity/cluster/ClusterLabel.java [hunk 4]
```diff
@@ -18,14 +18,17 @@
...
+import org.apache.linkis.manager.label.exception.LabelErrorException;
+
+import org.apache.commons.lang3.StringUtils;
 
 import java.util.HashMap;
 
-public class ClusterLabel extends GenericLabel implements EMNodeLabel {
+import static org.apache.linkis.manager.label.errorcode.LabelCommonErrorCodeSummary.LABEL_ERROR_CODE;
+
+public class ClusterLabel extends GenericLabel implements EngineNodeLabel, UserModifiable {
```
Comment (guoshupei) [inline linkis-computation-governance/linkis-manager/linkis-label-common/src/main/java/org/apache/linkis/manager/label/entity/cluster/ClusterLabel.java:31]: The column `label_key` of `linkis_cg_manager_label` needs to be extended to 50

File: linkis-computation-governance/linkis-manager/linkis-manager-common/src/main/java/org/apache/linkis/manager/common/entity/resource/KubernetesResource.java [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+      this.namespace = "default";
+    } else {
+      this.memory = Long.MAX_VALUE;
+      this.cores = Long.MAX_VALUE;
+      this.namespace = "default";
+    }
+  }
+
+  public KubernetesResource() {
+    this(Long.MAX_VALUE, Long.MAX_VALUE);
```
Comment (guoshupei) [inline linkis-computation-governance/linkis-manager/linkis-manager-common/src/main/java/org/apache/linkis/manager/common/entity/resource/KubernetesResource.java:54]: add default value of namespace

File: linkis-engineconn-plugins/spark/src/main/scala/org/apache/linkis/engineplugin/spark/factory/SparkEngineConnResourceFactory.scala [hunk 6]
```diff
@@ -20,19 +20,50 @@ package org.apache.linkis.engineplugin.spark.factory
...
 
+import io.fabric8.kubernetes.api.model.Quantity
+
 class SparkEngineConnResourceFactory extends AbstractEngineResourceFactory with Logging {
 
+  override protected def getMinRequestResource(
+      engineResourceRequest: EngineResourceRequest
+  ): Resource = {
+    val clusterLabel = LabelUtil.getLabelFromList[ClusterLabel](engineResourceRequest.labels)
+    if (clusterLabel != null && clusterLabel.getClusterType == "K8S") {
```
Comment (guoshupei) [inline linkis-engineconn-plugins/spark/src/main/scala/org/apache/linkis/engineplugin/spark/factory/SparkEngineConnResourceFactory.scala:49]: change '==' to `equals`， and use `toUpperCase` to be compatible with string matching

File: linkis-engineconn-plugins/spark/src/main/scala/org/apache/linkis/engineplugin/spark/factory/SparkEngineConnResourceFactory.scala [hunk 7]
```diff
@@ -20,19 +20,50 @@ package org.apache.linkis.engineplugin.spark.factory
...
+    } else {
+      getRequestResource(engineResourceRequest.properties)
+    }
+  }
+
+  override protected def getMaxRequestResource(
+      engineResourceRequest: EngineResourceRequest
+  ): Resource = {
+    val clusterLabel = LabelUtil.getLabelFromList[ClusterLabel](engineResourceRequest.labels)
+    if (clusterLabel != null && clusterLabel.getStringValue.startsWith("K8S")) {
```
Comment (guoshupei) [inline linkis-engineconn-plugins/spark/src/main/scala/org/apache/linkis/engineplugin/spark/factory/SparkEngineConnResourceFactory.scala:60]: use `toUpperCase` to be compatible with string matching

File: linkis-engineconn-plugins/spark/src/main/scala/org/apache/linkis/engineplugin/spark/factory/SparkEngineConnResourceFactory.scala [hunk 8]
```diff
@@ -63,4 +94,44 @@ class SparkEngineConnResourceFactory extends AbstractEngineResourceFactory with
...
+    val driverMemoryWithUnit = if (StringUtils.isNumeric(driverMemory)) {
+      driverMemory + "g"
+    } else {
+      driverMemory
+    }
+    val totalExecutorMemory = ByteTimeUtils.byteStringAsBytes(
+      executorMemoryWithUnit
+    ) * executorNum + ByteTimeUtils.byteStringAsBytes(driverMemoryWithUnit)
+    val totalExecutorCores = executorCores * executorNum + driverCores
+    // logger.info(s"总需求内存:$totalExecutorMemory, 总需求核数$totalExecutorCores")
```
Comment (guoshupei) [inline linkis-engineconn-plugins/spark/src/main/scala/org/apache/linkis/engineplugin/spark/factory/SparkEngineConnResourceFactory.scala:129]: Use English

File: linkis-computation-governance/linkis-manager/linkis-application-manager/src/main/java/org/apache/linkis/manager/rm/service/RequestResourceService.java [hunk 9]
```diff
@@ -404,7 +403,61 @@ public Pair<Integer, String> generateNotEnoughMessage(
...
+        return Pair.of(
+            RMErrorCode.NAMESPACE_CPU_INSUFFICIENT.getErrorCode(),
+            RMErrorCode.NAMESPACE_CPU_INSUFFICIENT.getErrorDesc()
+                + RMUtils.getResourceInfoMsg(
+                    RMConstant.CPU,
+                    RMConstant.KUBERNETES_CPU_UNIT,
+                    kubernetesResource.getCores(),
+                    kubernetesResourceAvailable.getCores(),
+                    kubernetesResourceMax.getCores()));
+      } else if (kubernetesResource.getMemory() > kubernetesResource.getMemory()) {
```
Comment (guoshupei) [inline linkis-computation-governance/linkis-manager/linkis-application-manager/src/main/java/org/apache/linkis/manager/rm/service/RequestResourceService.java:420]: change to "kubernetesResource.getMemory() > kubernetesResourceAvailable.getMemory()"

**Review by guoshupei — CHANGES_REQUESTED** [review]

**Review by guoshupei — COMMENTED** [review]
LGTM

File: linkis-computation-governance/linkis-manager/linkis-application-manager/pom.xml [hunk 10]
```diff
@@ -104,7 +104,31 @@
...
+    </dependency>
+    <dependency>
+      <groupId>io.fabric8</groupId>
+      <artifactId>kubernetes-model-common</artifactId>
+      <version>${kubernetes-client.version}</version>
+    </dependency>
+    <dependency>
+      <groupId>io.fabric8</groupId>
+      <artifactId>kubernetes-model-core</artifactId>
+      <version>5.4.1</version>
```
Comment (peacewong) [inline linkis-computation-governance/linkis-manager/linkis-application-manager/pom.xml:130]: can use props value?

File: linkis-dist/package/db/upgrade/1.4.0_schema/mysql/linkis_ddl.sql [hunk 11]
```diff
@@ -17,4 +17,5 @@
 
 ALTER TABLE `linkis_cg_manager_service_instance` ADD COLUMN `identifier` varchar(32) COLLATE utf8_bin DEFAULT NULL;
 ALTER TABLE `linkis_cg_manager_service_instance` ADD COLUMN `ticketId` varchar(255) COLLATE utf8_bin DEFAULT NULL;
-ALTER TABLE `linkis_cg_ec_resource_info_record` MODIFY COLUMN metrics TEXT DEFAULT NULL COMMENT 'ec metrics';
\ No newline at end of file
+ALTER TABLE `linkis_cg_ec_resource_info_record` MODIFY COLUMN metrics TEXT DEFAULT NULL COMMENT 'ec metrics';
+ALTER TABLE `linkis_cg_manager_label` MODIFY COLUMN label_key varchar(50);
```
Comment (peacewong) [inline linkis-dist/package/db/upgrade/1.4.0_schema/mysql/linkis_ddl.sql:21]: need to move 1.5.0_schema

File: linkis-engineconn-plugins/spark/pom.xml [hunk 12]
```diff
@@ -418,7 +418,6 @@
       <artifactId>kubernetes-model-core</artifactId>
       <version>5.4.1</version>
```
Comment (peacewong) [inline linkis-engineconn-plugins/spark/pom.xml:419]: can use props value?

File: linkis-engineconn-plugins/spark/src/main/scala/org/apache/linkis/engineplugin/spark/config/SparkResourceConfiguration.scala [hunk 13]
```diff
@@ -31,4 +31,22 @@ object SparkResourceConfiguration {
   val LINKIS_SPARK_EXECUTOR_INSTANCES = CommonVars[Int]("spark.executor.instances", 3)
   val LINKIS_QUEUE_NAME = CommonVars[String]("wds.linkis.rm.yarnqueue", "default")
 
+  val LINKIS_SPARK_KUBERNETES_NAMESPACE =
+    CommonVars[String]("spark.kubernetes.namespace", "default")
```
Comment (peacewong) [inline linkis-engineconn-plugins/spark/src/main/scala/org/apache/linkis/engineplugin/spark/config/SparkResourceConfiguration.scala:35]: need to be consistent
image

File: linkis-engineconn-plugins/spark/src/main/scala/org/apache/linkis/engineplugin/spark/factory/SparkEngineConnResourceFactory.scala [hunk 14]
```diff
@@ -20,19 +20,54 @@ package org.apache.linkis.engineplugin.spark.factory
...
+import io.fabric8.kubernetes.api.model.Quantity
+
 class SparkEngineConnResourceFactory extends AbstractEngineResourceFactory with Logging {
 
+  override protected def getMinRequestResource(
+      engineResourceRequest: EngineResourceRequest
+  ): Resource = {
+    val clusterLabel = LabelUtil.getLabelFromList[ClusterLabel](engineResourceRequest.labels)
+    if (
+        clusterLabel != null && StringUtils.equals(clusterLabel.getClusterType.toUpperCase(), "K8S")
```
Comment (peacewong) [inline linkis-engineconn-plugins/spark/src/main/scala/org/apache/linkis/engineplugin/spark/factory/SparkEngineConnResourceFactory.scala:50]: "k8s" should defined constant

**Review by peacewong — DISMISSED** [review]
LGTM.

**Review by peacewong — APPROVED** [review]
LGTM.

**Pull request merged**